File: 74-search-a-2d-matrix/search-a-2d-matrix.ts
function searchMatrix(matrix: number[][], target: number): boolean {
    const rows:number =  matrix.length;
    const cols: number = matrix[0].length;
    
    let left: number  = 0;
    let right: number = rows* cols - 1;
    
    while (left <= right){
        let midindex: number = Math.floor((left + right) / 2);
        let row: number = Math.floor(midindex / cols);
        let col: number = Math.floor(midindex % cols);
        
        const midvalue:number = matrix[row][col];
        
        if (midvalue === target){
            return true;
        }
        
        if (midvalue < target){
            left = midindex + 1;
        }
        
        else{
            right = midindex - 1;
        }
    
        
    }
    return false;   
};

File: 74-search-a-2d-matrix/search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        rows = len(matrix)
        cols = len(matrix[0])

        left = 0 
        right = rows * cols - 1 

        while left <= right:
            midindex = (left + right)  // 2
            row = midindex // cols
            col = midindex % cols

            mid_value = matrix[row][col]

            if mid_value == target:
                return True

            if mid_value < target:
                left = midindex + 1

            if mid_value > target:
                right = midindex - 1

        return False 
        
